src/lib.rs

```diff
@@ -265,11 +254,18 @@ impl NixPath for [u8] {
     where
         F: FnOnce(&CStr) -> T,
     {
-        if self.len() >= PATH_MAX as usize {
-            return Err(Errno::ENAMETOOLONG);
+        // The real PATH_MAX is 4096, but it's statistically unlikely to have a path longer than
+        // ~300 bytes. See the appendix to get stats for your own machine.
+        //
+        // By being smaller than a memory page, we also avoid the compiler inserting a probe frame:
+        // https://docs.rs/compiler_builtins/latest/compiler_builtins/probestack/index.html
+        const MAX_STACK_ALLOCATION: usize = 512;
```

Comment:
Was there a particular reason you chose `512`? Choosing the "best" value for `MAX_STACK_ALLOCATION` has been bothering me.

Considerations include things like:
- Avoiding `__rust_probestack` for the "common" case.
  - Only supported on x86 and x86_64 today, so let's assume a page is at least 4096 bytes.
  - Our stack usage is dependent on the size of `T`, but excluding `buf` is typically fairly low.
- Avoiding heap allocations where possible.
  - Only fall back to `with_nix_path_allocating` where it is necessary or otherwise justifiable.
- Avoiding using significantly more stack space than previously.
  - `PATH_MAX` is 1024 on the BSDs, so indiscriminately picking a much larger value isn't appropriate.

Choosing `4032` (as I believe you had done previously?) seems to mostly alleviate the first two points, but is significantly higher than the current stack utilization on the BSDs. On the other hand, choosing `512` addresses the first and third points, but still leads to heap allocations on all platforms for valid path lengths.

IMO `1024` seems like a reasonable compromise. It should fully avoid `__rust_probestack` for sane values of `T`. It ensures that we don't get a heap allocation on the BSDs while not increasing the stack usage, and also ensures we won't get a heap allocation on other platforms for reasonable path lengths. We could optionally have different definitions per-platform, but that feels like we're getting into diminishing returns.

Reply:
Yeah, 512 was pretty arbitrary. I bumped it to 1024, but I'm even happy to go to 2048 if you think that's worth it. Might be interesting to run that histogram script on your machine to see what your longest path is.

Reply:
@asomers I'm curious for your thoughts on the maximum stack allocation size here.

Reply:
There really isn't any magic number for what's acceptable for the stack.  I don't have any better idea that Supercilex does.